dask/array/utils.py

```diff
@@ -77,6 +77,38 @@ def meta_from_array(x, ndim=None, dtype=None):
     return meta
 
 
+def compute_meta(func, dtype, *args, **kwargs):
+    args_meta = list(map(meta_from_array, args))
+    kwargs_meta = {k: meta_from_array(v) for k, v in kwargs.items()}
+
+    # todo: look for alternative to this, causes issues when using map_blocks()
+    # with np.vectorize, such as dask.array.routines._isnonzero_vec().
+    if isinstance(func, np.vectorize):
+        meta = func(*args_meta)
+        if hasattr(meta, 'dtype'):
+            return meta.astype(dtype)
+        else:
+            return None
+
+    try:
+        meta = func(*args_meta, **kwargs_meta)
+    except TypeError as e:
+        if ("unexpected keyword argument" in str(e) or
+                "is an invalid keyword for" in str(e)):
+            raise
+        else:
+            return None
+    except Exception:
+        return None
+
+    if hasattr(meta, 'dtype'):
+        meta = meta.astype(dtype)
```

Comment:
If `dtype=None`, will that have no effect or may it return something unexpected?

Reply:
We now only call astype if a dtype is specified